### PR TITLE
`Integrated code lifecycle`: Add finished build jobs view 

### PR DESCRIPTION
**Pull request by mateusmm01**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).

### Motivation and Context

Currently we don't have a view to see all finished build jobs.

### Description

A finished build job table was added in the BuildQueueView. This was implemented using pagination. We only get 50 jobs at a time and fetch these ordered by completion date descending to first get the most recent build jobs.

### Steps for Testing

Prerequisites:
- 1 Admin
- 1 Programming Exercise 

1. Make some submissions and wait until they are done running 
2. Navigate to Course Management > Build Queue and check that you can see the build job in the finished build jobs table
3. Navigate to System Administration > Build Queue and check that you can see the build job here too but also build jobs from other courses

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [build-job.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6451/latest/artifact/shared/Coverage-Report-Client-Tests/app/entities/build-job.model.ts.html) | 100% | ✅ |
| [build-queue.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6451/latest/artifact/shared/Coverage-Report-Client-Tests/app/localci/build-queue/build-queue.component.ts.html) | 88.78% | ✅  |
| [build-queue.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6451/latest/artifact/shared/Coverage-Report-Client-Tests/app/localci/build-queue/build-queue.service.ts.html) | 100% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [AdminBuildJobQueueResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6451/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.admin/AdminBuildJobQueueResource.html) | 100% | ✅ |
| [BuildJobQueueResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6451/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.localci/BuildJobQueueResource.html) | 91% | ✅  |
| [PageUtil.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6451/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.util/PageUtil.html) | 91% | ✅  |

### Screenshots

image

## Summary by CodeRabbit

- **New Features**
	- Added eager loading methods for retrieving build jobs with pagination support.
	- Improved URL routing consistency in Aeolus Template API.
	- Introduced API method to retrieve finished build jobs for specific courses.
	- Updated German localization with new build job status terms.

- **Refactor**
	- Enhanced internal handling of build job data retrieval and pagination.

- **Style**
	- Adjusted API endpoint paths to improve consistency.

- **Documentation**
	- Updated module imports documentation in course management features.

### CODE REVIEW

**Review by rstief — APPROVED** [review]
Demonstrated in testing session, looks good

**Review by florian-glombik — APPROVED** [review]
Demonstrated on ts3, worked as described 👍

**Review by rstief — DISMISSED** [review]

File: src/main/java/de/tum/in/www1/artemis/repository/BuildJobRepository.java [hunk 1]
```diff
@@ -29,4 +32,6 @@ public interface BuildJobRepository extends JpaRepository<BuildJob, Long> {
             """)
     Set<DockerImageBuild> findAllLastBuildDatesForDockerImages();
 
+    Page<BuildJob> findAllByCourseId(Long courseId, Pageable pageable);
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/repository/BuildJobRepository.java:35]:
```suggestion
    Page<BuildJob> findAllByCourseId(long courseId, Pageable pageable);
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/localci/BuildJobQueueResource.java [hunk 2]
```diff
@@ -133,4 +146,24 @@ public ResponseEntity<Void> cancelAllRunningBuildJobs(@PathVariable long courseI
...
+     */
+    @GetMapping("/courses/{courseId}/finished-jobs")
+    @EnforceAtLeastInstructor
+    public ResponseEntity<List<BuildJob>> getFinishedBuildJobsForCourse(@PathVariable long courseId, PageableSearchDTO<String> search) {
+        log.debug("REST request to get the finished build jobs for course {}", courseId);
+        Course course = courseRepository.findByIdElseThrow(courseId);
+        if (!authorizationCheckService.isAtLeastInstructorInCourse(course, null)) {
+            throw new AccessForbiddenException("You are not allowed to access finished build jobs of this course!");
+        }
+        final Page<BuildJob> page = buildJobRepository.findAllByCourseId(courseId, PageUtil.createDefaultPageRequest(search, PageUtil.ColumnMapping.BUILD_JOB));
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/web/rest/localci/BuildJobQueueResource.java:164]:
```suggestion
    @EnforceAtLeastInstructorInCourse
    public ResponseEntity<List<BuildJob>> getFinishedBuildJobsForCourse(@PathVariable long courseId, PageableSearchDTO<String> search) {
        log.debug("REST request to get the finished build jobs for course {}", courseId);
        final Page<BuildJob> page = buildJobRepository.findAllByCourseId(courseId, PageUtil.createDefaultPageRequest(search, PageUtil.ColumnMapping.BUILD_JOB));
```

File: src/main/webapp/i18n/de/buildQueue.json [hunk 3]
```diff
@@ -5,6 +5,7 @@
             "pageTitle": "Build Job Informationen",
             "queuedBuildJobs": "Wartende Build Jobs",
             "runningBuildJobs": "Laufende Build Jobs",
+            "finishedBuildJobs": "Fertige Build Jobs",
```
Comment (Strohgelaender) [inline src/main/webapp/i18n/de/buildQueue.json:8]:
```suggestion
            "finishedBuildJobs": "Abgeschlossene Build Jobs",
```

**Review by rstief — DISMISSED** [review]
reapprove

**Review by laurenzfb — DISMISSED** [review]
server code lgtm, please solve merge conflicts

File: src/main/java/de/tum/in/www1/artemis/repository/BuildJobRepository.java [hunk 4]
```diff
@@ -19,12 +23,15 @@
...
 @Repository
-public interface BuildJobRepository extends JpaRepository<BuildJob, Long> {
+public interface BuildJobRepository extends JpaRepository<BuildJob, Long>, JpaSpecificationExecutor<BuildJob> {
 
     Optional<BuildJob> findFirstByParticipationIdOrderByBuildStartDateDesc(Long participationId);
 
     Optional<BuildJob> findBuildJobByResult(Result result);
 
+    @EntityGraph(attributePaths = { "result", "result.feedbacks", "result.feedbacks.testCase", "result.participation", "result.submission", "result.participation.exercise" })
+    Page<BuildJob> findAll(Pageable pageable);
```
[the remark this answers is not in the document]
Comment (laurenzfb) [reply src/main/java/de/tum/in/www1/artemis/repository/BuildJobRepository.java:33]: You are right, I will have to refactor this

**Review by rstief — DISMISSED** [review]
tested on ts11, works as expected.

File: src/main/java/de/tum/in/www1/artemis/repository/BuildJobRepository.java [hunk 5]
```diff
@@ -19,12 +23,15 @@
...
 @Profile(PROFILE_CORE)
 @Repository
-public interface BuildJobRepository extends JpaRepository<BuildJob, Long> {
+public interface BuildJobRepository extends JpaRepository<BuildJob, Long>, JpaSpecificationExecutor<BuildJob> {
 
     Optional<BuildJob> findFirstByParticipationIdOrderByBuildStartDateDesc(Long participationId);
 
     Optional<BuildJob> findBuildJobByResult(Result result);
 
+    @EntityGraph(attributePaths = { "result", "result.participation", "result.participation", "result.participation.exercise", "result.submission" })
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/repository/BuildJobRepository.java:32]:
```suggestion
    @EntityGraph(attributePaths = { "result", "result.participation", "result.participation.exercise", "result.submission" })
```

**Review by JohannesWt — DISMISSED** [review]
Retested on staging - still works as expected

**Review by rstief — APPROVED** [review]

**Pull request merged**